wnfs-common/src/blockstore.rs

```diff
@@ -8,134 +6,181 @@ use libipld::{
+    let second_bytes = b"hello world".to_vec();
 
-        self.0.insert(cid.to_string(), bytes);
+    // Insert the objects into the blockstore
+    let first_cid = store.put_serializable(&first_bytes).await.unwrap();
+    let second_cid = store.put_serializable(&second_bytes).await.unwrap();
 
-        Ok(cid)
-    }
+    // Retrieve the objects from the blockstore
+    let first_loaded: Vec<u8> = store.get_deserializable(&first_cid).await.unwrap();
+    let second_loaded: Vec<u8> = store.get_deserializable(&second_cid).await.unwrap();
 
-    /// Retrieves an array of bytes from the block store with given CID.
-    async fn get_block<'a>(&'a self, cid: &Cid) -> Result<Cow<'a, Vec<u8>>> {
-        let bytes = self
-            .0
-            .get(&cid.to_string())
-            .ok_or(BlockStoreError::CIDNotFound(*cid))?;
+    // Assert that the objects are the same as the ones we inserted
+    assert_eq!(first_loaded, first_bytes);
+    assert_eq!(second_loaded, second_bytes);
 
-        Ok(Cow::Borrowed(bytes))
-    }
+    // Return Ok
+    Ok(())
 }
 
-//--------------------------------------------------------------------------------------------------
-// Functions
-//--------------------------------------------------------------------------------------------------
+// Generic function used to test any type that conforms to the BlockStore trait
+pub async fn bs_duplication<T: BlockStore + Send + 'static>(store: &T) -> Result<()> {
+    // Example objects to insert and remove from the blockstore
+    let first_bytes = vec![1, 2, 3, 4, 5];
+    let second_bytes = first_bytes.clone();
+
+    // Insert the objects into the blockstore
+    let first_cid = store.put_serializable(&first_bytes).await.unwrap();
+    let second_cid = store.put_serializable(&second_bytes).await.unwrap();
+
+    // Assert that the two vecs produced the same CID
+    assert_eq!(first_cid, second_cid);
+
+    // Retrieve the objects from the blockstore
+    let first_loaded: Vec<u8> = store.get_deserializable(&first_cid).await.unwrap();
+    let second_loaded: Vec<u8> = store.get_deserializable(&second_cid).await.unwrap();
+
+    // Assert that the objects are the same as the ones we inserted
+    assert_eq!(first_loaded, first_bytes);
+    assert_eq!(second_loaded, second_bytes);
+    // Assert that the objects we loaded are the same
+    assert_eq!(first_loaded, second_loaded);
+
+    // Return Ok
+    Ok(())
+}
+
+pub async fn bs_serialization<
+    T: BlockStore + Send + Serialize + 'static + for<'de> Deserialize<'de>,
+>(
+    store: &T,
+) -> Result<()> {
+    // Example objects to insert and remove from the blockstore
+    let bytes = vec![1, 2, 3, 4, 5];
+    // Insert the object into the blockstore
+    let cid = store.put_serializable(&bytes).await.unwrap();
+
+    // Create a buffer to hold the serialized BlockStore
+    let mut writer: Vec<u8> = Vec::new();
+    // Serialize the blockstore
+    serde_json::to_writer_pretty(&mut writer, &store).unwrap();
```

Comment:
Would you mind using the `dagcbor::encode` and `dagcbor::decode` functions from the `encodings.rs` file instead of `serde_json`? I'd like to avoid pulling in the serde_json dependencies for this :v:

Reply:
switching this out 👍🏻